### DISCUSSION TITLE
Considerations for adding cjpegli as potential jpeg encoder?

### DISCUSSION THREAD

**Discussion by Earnestly** [ocrmypdf/OCRmyPDF#1381] [Ideas]
The team who made jpegxl's `cjxl` have incorporated much of their discoveries into a tool called `cjpegli` for use with jpeg. https://github.com/libjxl/libjxl/blob/main/lib/jpegli/README.md

See also https://giannirosato.com/blog/post/jpegli/

From my personal testing a straight conversion with no arguments has roughly halved the size of my picture collection (although admittedly they were likely not very compressed to begin with).

Is it possible to add this encoder under `_exec` along with jpeg2000 and jbig2?

**Comment by jbarlow83** [comment]
Could be done with a plugin, but I can't see taking this on myself any time soon.

The basic implementation is straightforward, but some of the ways that program reformats JPEGs would need some adjustment to work with PDF and testing time to confirm it works with all PDF viewers.

Perhaps the authors of cjpegli could be inspired to contribute their changes to the Python Pillow project. Then all Python software that works with JPEG would benefit.

**Comment by Earnestly** [comment]
Yes, the pillow point would be pretty wonderful. There doesn't yet seem to be much discussion within pillow about adding it.

Also just to be clear about this, `cjpegli` produces ordinary jpegs. It's not incompatible with readers or viewers that can already decode jpeg. Like `mozjpeg`, it's a drop-in replacement encoder.
